### ISSUE TITLE
"outer" join

### ISSUE THREAD

**Issue by jmatsushita** [beautifulinteractions/node-quadstore#48]
Hi,

It would be nice to have an "outer" join such that this code:
```
const QuadStore = require('quadstore').QuadStore;
const store = new QuadStore('./db', { db: require('leveldown') });

const initialQuads = [
  { subject: 'g0s', predicate: 'g0p', object: 'g0o', graph: 'g1s' },
  { subject: 'g1s', predicate: 'g1p', object: 'g1o', graph: 'g2s' },
  { subject: 'g2s', predicate: 'g2p', object: 'g2o', graph: 'g3s' }
];

store
  .del({}).then(() => {
    store
      .put(initialQuads)
      .then(() => {
        return store
          .query({})
          .join(store.query({}), ['graph'], ['subject'])
          .get()
          .then((quads) => console.log(format(quads)));
      })
    })

const format = (quads) => quads
  .map(({subject, predicate, object, graph}) => ({s: subject, p: predicate, o: object, g: graph}))
  .reduce((acc, {s,p,o,g}) => acc + `${s}\t${p}\t${o}\t${g}` + "\n", "\n")
```

Instead of outputting currently:
```
g0s	g0p	g0o	g1s
g1s	g1p	g1o	g2s
```

Would output the third triple too (matching from queryB). i.e.
```
g0s	g0p	g0o	g1s
g1s	g1p	g1o	g2s
g2s	g2p	g2o	g3s
```

Cheers,

Jun

**Comment by jacoscaz** [comment]
This should be reasonably easy to implement by adding a `outerJoin` method which modifies the behaviour of the `JoinStream` object by allowing all triples from the first query to go through.

Something like:

```
store.query({})
  .outerJoin(store.query({}), ['graph'], ['subject'])
```

**Comment by jacoscaz** [comment]
@jmatsushita 

Sorting this out is harder than I thought but I *think* I've got it working here: https://github.com/jacoscaz/node-quadstore/tree/outer-join .

Would you be able to test this?

**Comment by jmatsushita** [comment]
Hi @jacoscaz 

I just gave it a try with the code above and it doesn't terminate.

I added some debug code and it seems that your `_drainA` function just keeps draining even though the `quad` is `null`.

How is `keepDraining` set to false [here](https://github.com/jacoscaz/node-quadstore/blob/1b73e4c36567762df91ad0a4743298cfd1d8aeb6/lib/query/outer-join-query.js#L41-L49)?

Jun

**Comment by jacoscaz** [comment]
Hi @jmatsushita 

Thank you for your help with testing this. That should now be fixed at https://github.com/jacoscaz/node-quadstore/tree/outer-join. I've also fixed a few bugs related tuples being left out of joins using partial term names (i.e. looking at less terms than the full set of 4 terms). This needs a ton more unit tests to cover all edge cases but, if you can spare the time, some hands-on testing on your part would be highly appreciated.

**Comment by jmatsushita** [comment]
Hi @jacoscaz 

Just pulled and relaunched and saw this with the above code:
```
g0s	g0p	g0o	g1s
g1s	g1p	g1o	g2s
g1s	g1p	g1o	g2s
g2s	g2p	g2o	g3s
g2s	g2p	g2o	g3s
```

So I think this is outputting duplicates...

Cheers,

Jun

**Comment by jacoscaz** [comment]
Hello @jmatsushita . Sorry for the wait.

While working on this issue we unearthed a few other issues with the general join mechanism and ended up rewriting both the `innerJoin` and `outerJoin` logic from the start.

In addition to this, we've thought a bit about what a `join`, an `innerJoin` and a `outerJoin` should be. Even though the current release behaves as described in the `README`

> Performs an inner join between the two queries limited to the terms specified in the compareTerms arrays, passing on quads from the *first* query.

we're thinking a `join` should really pass on quads from *both* queries to maintain a parallel with traditional `JOIN` statements in relational databases, where columns from both queries appear in the results. At the moment, we feel like we're conflating a `join` with a `filter`.

My branch at https://github.com/jacoscaz/node-quadstore/tree/outer-join now reflects all of these changes. 

Going back to your latest feedback... I've included your case in the mocha test suite and duplication seems to be gone now. That said, because of the new approach to `join`s, using your specific code both the `innerJoin()` and the `outerJoin()` methods should return the entire set of triples.
